src/3d/qgsrulebased3drenderer.cpp

```diff
@@ -0,0 +1,470 @@
+  Date                 : January 2019
+  Copyright            : (C) 2019 by Martin Dobias
+  Email                : wonder dot sk at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsrulebased3drenderer.h"
+
+#include "qgsvectorlayer.h"
+#include "qgsxmlutils.h"
+
+#include "qgs3dmapsettings.h"
+#include "qgsline3dsymbol.h"
+#include "qgspoint3dsymbol.h"
+#include "qgspolygon3dsymbol.h"
+#include "qgsline3dsymbol_p.h"
+#include "qgspoint3dsymbol_p.h"
+#include "qgspolygon3dsymbol_p.h"
+
+QgsRuleBased3DRendererMetadata::QgsRuleBased3DRendererMetadata()
+  : Qgs3DRendererAbstractMetadata( QStringLiteral( "rulebased" ) )
+{
+}
+
+QgsAbstract3DRenderer *QgsRuleBased3DRendererMetadata::createRenderer( QDomElement &elem, const QgsReadWriteContext &context )
+{
+  QDomElement rulesElem = elem.firstChildElement( QStringLiteral( "rules" ) );
+
+  QgsRuleBased3DRenderer::Rule *root = QgsRuleBased3DRenderer::Rule::create( rulesElem, context );
+  if ( !root )
+    return nullptr;
+
+  QgsRuleBased3DRenderer *r = new QgsRuleBased3DRenderer( root );
+  r->readXml( elem, context );
+  return r;
+}
+
+
+// ---------
+
+
+////////////////////
+
+QgsRuleBased3DRenderer::Rule::Rule( QgsAbstract3DSymbol *symbol, const QString &filterExp, const QString &description, bool elseRule )
+  : mSymbol( symbol )
+  , mFilterExp( filterExp )
+  , mDescription( description )
+  , mElseRule( elseRule )
+
+{
+  initFilter();
+}
+
+QgsRuleBased3DRenderer::Rule::~Rule()
+{
+  qDeleteAll( mChildren );
+  // do NOT delete parent
+}
+
+void QgsRuleBased3DRenderer::Rule::setSymbol( QgsAbstract3DSymbol *symbol )
+{
+  if ( mSymbol.get() == symbol )
+    return;
+
+  mSymbol.reset( symbol );
+}
+
+QgsRuleBased3DRenderer::RuleList QgsRuleBased3DRenderer::Rule::descendants() const
+{
+  RuleList l;
+  for ( Rule *c : mChildren )
+  {
+    l += c;
+    l += c->descendants();
+  }
+  return l;
+}
+
+void QgsRuleBased3DRenderer::Rule::initFilter()
+{
+  if ( mElseRule || mFilterExp.compare( QLatin1String( "ELSE" ), Qt::CaseInsensitive ) == 0 )
+  {
+    mElseRule = true;
+    mFilter.reset( nullptr );
+  }
+  else if ( !mFilterExp.isEmpty() )
+  {
+    mFilter.reset( new QgsExpression( mFilterExp ) );
+  }
+  else
+  {
+    mFilter.reset( nullptr );
+  }
+}
+
+void QgsRuleBased3DRenderer::Rule::updateElseRules()
+{
+  mElseRules.clear();
+  Q_FOREACH ( Rule *rule, mChildren )
```

Comment:
Q_FOREACH? waaaaaat? :stuck_out_tongue:

Reply:
a small trap to check if reviewers are paying attention ;-) fixed in a01caa43